plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -496,11 +498,14 @@ private static ConnectorPageSource createParquetPageSource(
             Predicate parquetPredicate = buildPredicate(requestedSchema, parquetTupleDomain, descriptorsByPath, UTC);
 
             List<BlockMetaData> blocks = new ArrayList<>();
+            ImmutableList.Builder<Optional<ColumnIndexStore>> columnIndexes = ImmutableList.builder();
             for (BlockMetaData block : parquetMetadata.getBlocks()) {
                 long firstDataPage = block.getColumns().get(0).getFirstDataPageOffset();
+                Optional<ColumnIndexStore> columnIndex = getColumnIndexStore(dataSource, block, descriptorsByPath, parquetTupleDomain, options);
```

Comment:
May be worth putting off the call to `getColumnIndexStore` utnil after the `start <= firstDataPage && firstDataPage < start + length` checks.

Reply:
Although we won't read the column index from file until later, it is better to avoid creating the object until it's needed. We can make same change in ParquetPageSourceFactory as well.